src/transactions/cpfp.rs

```diff
@@ -0,0 +1,209 @@
+    error::*,
+    transactions::{utils, CpfpableTransaction, RevaultTransaction, CPFP_MIN_CHANGE},
+    txins::*,
+    txouts::*,
+};
+
+use miniscript::bitcoin::{
+    consensus::encode::Decodable,
+    util::psbt::{Input as PsbtIn, PartiallySignedTransaction as Psbt},
+    Amount, Script, SigHashType, Transaction, TxIn,
+};
+
+#[cfg(feature = "use-serde")]
+use {
+    serde::de::{self, Deserialize, Deserializer},
+    serde::ser::{Serialize, Serializer},
+};
+
+impl_revault_transaction!(
+    CpfpTransaction,
+    doc = "The transaction feebumping either an unvault or a spend"
+);
+
+const OP_RETURN_DUMMY_DATA: [u8; 22] = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+
+impl CpfpTransaction {
+    /// Create a new cpfp transaction from the transaction that needs cpfp, its fees and weight;
+    /// the required feerate (expressed in msats/WU); a list of CpfpTxIn that can be used along
+    /// with their child number, the cpfp descriptor and the secp curve.
+    /// The transaction that needs CPFP must be finalized, revault_tx will panic otherwise
+    pub(crate) fn from_tx(
+        to_be_cpfped: impl CpfpableTransaction,
+        feerate: u64,
+        mut available_utxos: Vec<CpfpTxIn>,
+    ) -> Result<CpfpTransaction, TransactionCreationError> {
+        assert!(to_be_cpfped.is_finalized());
+        let to_be_cpfped_txid = to_be_cpfped.txid();
+        // Checking that the feerate requested is higher than the previous
+        // tx feerate (otherwise why cpfping in the first place?)
+        let to_be_cpfped_fees = Amount::from_sat(to_be_cpfped.fees());
+        let to_be_cpfped_weight = to_be_cpfped.into_tx().get_weight();
+
+        // Math is hard
+        // to_be_cpfped_fees / to_be_cpfped_weight >= feerate / 1000
+        // (feerate is msats/WU, feerate / 1000 is sats/WU)
+        // Multiplying on both sides by to_be_cpfped_weight, 1000:
+        if 1000 * to_be_cpfped_fees.as_sat() >= feerate * to_be_cpfped_weight as u64 {
+            return Err(TransactionCreationError::FeerateTooLow);
+        }
+
+        // This will sort the vector in ascending order.
+        // Since we're going to pop() from it, we're using a largest first CS.
+        available_utxos.sort_unstable_by_key(|l| l.txout().txout().value);
+
+        // Let's pick the input to be cpfped
+        let to_be_cpfped_index = available_utxos
+            .iter()
+            .position(|l| l.outpoint().txid == to_be_cpfped_txid)
+            .ok_or(TransactionCreationError::MissingCpfpTxIn)?;
+
+        // Pushing it at the end of the vector, so it gets picked first
+        let u = available_utxos.remove(to_be_cpfped_index);
+        let dummy_change = u.txout().txout().clone();
+        available_utxos.push(u);
+
+        let mut inputs_sum = Amount::from_sat(0);
+        let mut total_satisfation_weight = 0;
+
+        let transaction = Transaction {
+            version: 2,
+            lock_time: 0,
+            input: vec![],
```

Comment:
Yep, I'm adding it in the loop as I already know it will be picked first. (https://github.com/revault/revault_tx/pull/115/files/cfc5142c4213f4b207728055dbf57fbdca49b543#diff-d75ff403bf5cee245364f4e7970d2a49b21a5f24e3dddd61377198b6e1fa66f1R63)